### ISSUE TITLE
Multiplexed

### ISSUE THREAD

**Issue by d11wtq** [d11wtq/dockerpty-restored-2#96]
#2 Issue by __delfick__, created on: 2014-06-27T13:10:14Z

This continues on from my other pull request https://github.com/d11wtq/dockerpty/pull/1

Takes into account this thing docker does where we get 8 byte headers that we have to deal with.

(with thanks to https://github.com/dotcloud/docker-py/blob/a2f0f1b2ddbc166dcaa26ada67d94973aff7a9c4/docker/client.py#L255 for the inspiration)

**Comment by d11wtq** [comment]
Comment by __d11wtq__, created on: 2014-06-29T08:27:38Z

Looks like changes I've made upstream have broken this, sorry :)

I need to understand this code a bit better before I merge it. I'll get back to you ASAP.

Just curious, but does the multiplexing only come into play if you request output from stdout and stderr in the same stream?

**Comment by d11wtq** [comment]
Comment by __delfick__, created on: 2014-06-29T09:10:24Z

I rebased and force pushed this branch to include your changes.

I'll have a look at how the combination of stderr, stdin and stdout effects it

**Comment by d11wtq** [comment]
Comment by __delfick__, created on: 2014-06-29T09:15:22Z

It looks like if I only have stdout or only have stderr it is still multiplexed.

It seems as soon as I add stdin it is no longer multiplexed.

**Comment by d11wtq** [comment]
Comment by __d11wtq__, created on: 2014-06-29T09:18:05Z

Wow, that's surprising to me and seems like a strange thing to do. Maybe docker is just using the same approach when you separate the streams and when you request them together. I'll start adding test coverage on the non-interactive functionality! Thanks.

**Comment by d11wtq** [comment]
Comment by __delfick__, created on: 2014-06-29T10:44:11Z

Thanks

(and yes, I also thought it would only multiplex if it had both stdout and stderr)

**Comment by d11wtq** [comment]
Comment by __d11wtq__, created on: 2014-06-30T06:31:03Z

Ok, this has been merged into non-interactive, thanks! I'll hopefully get a release out this evening.

**Comment by d11wtq** [comment]
Comment by __delfick__, created on: 2014-06-30T06:36:29Z

Awesome, thanks

**Comment by d11wtq** [comment]
Comment by __d11wtq__, created on: 2014-06-30T09:44:58Z

__delfick__ I think there's a bug, as I've written some tests and the one with `stdin_open=True` but `tty=False` fails because the 8-byte header is there. Now the API documentation says multiplexing occurs in all cases where `tty=False`, so I changed the check to do that and it seems to incorrectly swallow 8 bytes of "Hello World!" and produces "rld!".

See: https://travis-ci.org/d11wtq/dockerpty/builds/28761614

Seems like a good time to start writing actual unit tests, as this logic is complicated enough to warrant it. I'd maybe even propose making `io.Demuxer` that wraps an IO object, (then you can just use a list comprehension of map to do the wrapping) and unit test this in isolation (traditional unit tests, not with behave). That said, if you're able to just hack a fix together in 5 mins, go ahead and update this branch and I'll cherry-pick your change it into non-interactive.

I wonder if you need to mux data you write to stdin too, not just demux it on the way out.

I'm writing behavioural tests for all the things that need to work anyway, then I'll get onto this if you haven't beaten me to it.

**Comment by d11wtq** [comment]
Comment by __delfick__, created on: 2014-06-30T10:20:17Z

mmkay, I'll have a look later when I have some time, and I'll start writing some unit tests.

So what are you saying you changed the test to do for it to only say "rld!"?

**Comment by d11wtq** [comment]
Comment by __d11wtq__, created on: 2014-06-30T10:35:46Z

__delfick__ no dramas, I'll have some bandwidth too (although my temporary 3G internet may be about to die!).

In order for it to output `rld!` I changed the line:

``` python
options = {"multiplexed": not info["Config"]["AttachStdin"]}
```

To:

``` python
options = {"multiplexed": not info["Config"]["Tty"]}
```

**Comment by d11wtq** [comment]
Comment by __d11wtq__, created on: 2014-06-30T11:00:28Z

Figured it out. That change above is needed, but also demuxing should *not* be done to `sys.stdin`. It was actually swallowing the data being written to stdin :) Fixed!

**Comment by d11wtq** [comment]
Comment by __d11wtq__, created on: 2014-06-30T11:08:39Z

Merged. Because I rebased one of your commits out (the KeyboardInterrupt) GitHub hasn't detected it, but I'll close this. Your changes are in master and passing the build :) Will get a release out in a jiffy!

Thanks again for your hard work!

**Comment by d11wtq** [comment]
Comment by __d11wtq__, created on: 2014-06-30T11:19:18Z

Released to PyPi in version 0.2.0.

**Comment by d11wtq** [comment]
Comment by __delfick__, created on: 2014-06-30T11:25:49Z

yay! and just as I get back to the computer.

Thanks for making things work!
